proto/substrait/algebra.proto

```diff
@@ -340,6 +340,23 @@ message ExchangeRel {
   }
 }
 
+// The expand relation generates multiple rows per input row (denoted by
+// aggregate_expressions) which are then aggregated by groupings. No
+// ordering is applied to the aggregated results.
+message ExpandRel {
+  RelCommon common = 1;
+  Rel input = 2;
+
+  // A list of expression grouping that the aggregation measured should be calculated for.
+  repeated Grouping groupings = 4;
+
+  message Grouping {
```

Comment:
where is `ExpandRel` output?

Reply:
Do we need to define the output of the physical operator? It seems the output is defined by the consumer.

Reply:
How does `Aggregate` 's  operator reference the output of `Expand`

Reply:
@baibaichen Added the `output ` in `ExpandRel`.

Reply:
I disagree with this.  Substrait relations do not typically include names for the output.  See the `ProjectRel` for an example.

Reply:
> I disagree with this. Substrait relations do not typically include names for the output. See the `ProjectRel` for an example.

@westonpace, you are right, the name is unnecessry,  `Type.Struct` should be fine.  We actually need the index (for oder) and the type (for nullability) of `ExpandRel`'s ouput.

Reply:
I believe the order should be the same as `ProjectRel`.  By default the newly created columns will be at the end.  You can use an emit relation to reorder them where needed.  Is there some reason this wouldn't work?

The type should come from the expressions themselves.  Every expression is has a return type and so you should be able to determine the output type of the columns from the expressions.  Again, this is the same way the `ProjectRel` works.

Reply:
IIUC, expression need refer to input from child plan. However, there is no such input for outupt of `ExpandRel`

Reply:
I'm not certain I understand.  Let us consider the example above:

```
The physical plan looks like Scan -> Expand(output, projections)-> Aggregate[groupby, measure]
```

The expressions in `groupings` will refer to the output schema of the scan in this case.  So, for example, to refer to the Country column:

```
{
  "selection": {
    "direct_reference": {
      "struct_field": 1 # Assuming 1 is the index of Country in the scan output
    },
    "root_reference": {}
  }
}
```

The type can be found by looking at the output schema of scan.  To insert a literal `NULL` the expression would be...

```
{
  "literal": {
    "null": {
      "string": {}
    }   
  }
}
```
All literals, including null literals, are typed (all expressions have a clearly defined type).

Reply:
@westonpace 

> I'm not certain I understand. Let us consider the example above:
> ```
> The physical plan looks like Scan -> Expand(output, projections)-> Aggregate[groupby, measure]
> ```
> The expressions in groupings will refer to the output schema of the scan in this case. So, for example, to refer to the Country column:

1. `groupings` of `ExpandRel` will refer to its child's output, for example, it refers to scan in your example.
2.  `output` of `ExpandRel` is the input of  `AggregateRel` which is `ExpandRel`'s parent.

The reason that we can't refer to `groupings' is that there is a group of expressions, and each group with different nullable property.

Spark mixes measure column and group column into `Expand`'s `projections` (which @JkSelf named `groups` in substrait's ExpandRel), and we can't distinguish them easily.​

Reply:
> If this is only duplicating records (and nothing else

It isn't duplication and nothing else though.  Each copy of the input row will have a unique projection applied to it.  Consider the above example.  The groupings / projection expressions will be:

```
[
  [field("country"), field("region")]
  [field("country"), literal(null, string())]
  [literal(null, string()), literal(null, string())]
]
```

The grouping id (like in aggregate rel) can be implicit (not specified in the projections).  Given an input row:

| Country | Region | Sales |
| --- | --- | --- |
| US | North America | 100 |

There will be three output rows.  Before any emit they will be:

| Country | Region | Sales | Expr0 | Expr1 | Grouping Id |
| --- | --- | --- | --- | --- | --- |
| US | North America | 100 | US | North America | 0 |
| US | North America | 100 | US | NULL | 1 |
| US | North America | 100 | NULL | NULL | 3 |

Presumably, an emit will be added `[2, 3, 4, 5]` to yield:

| Sales | Expr0 | Expr1 | Grouping Id |
| --- | --- | --- | --- |
| 100 | US | North America | 0 |
| 100 | US | NULL | 1 |
| 100 | NULL | NULL | 3 |

Reply:
> I don't understand why the same behavior as aggregate wouldn't work here. It may not be how people have thought about it but it seems like it would make things consistent and would apply cleanly onto the concepts trying to be capture here.

I think that is fine.  This particular node operates more similarly to projection than to aggregation in my mind.  So I was suggesting we keep the output order similar to projection.

However, I don't know of any case where the original columns, once referenced, wouldn't be dropped.  So if we want the output order to be:

```
unreferenced columns | grouping expressions | grouping id
```

I think that should still work (though "unreferenced columns" is a bit of a new concept).

> My bad, you're right that it's duplication + nullification of specific columns per grouping, (same as the already specified behavior in aggregate).

Yes.  I think it would be useful if the `nullification` wasn't required though.  The operator would be simpler to understand and then this expand behavior is also useful for [`pivot_longer`](https://tidyr.tidyverse.org/reference/pivot_longer.html)

Given an input table:

| School | Grade 1 Avg Score | Grade 2 Avg Score |
| --- | --- | --- |
| Alpha Elementary | 95 | 92 |
| Beta Elementary | 97 | 93 |

and an expand spec:

```
[
  [literal("Grade 1", string()), field("Grade 1 Avg Score")],
  [literal("Grade 2", string()), field("Grade 2 Avg Score")]
]
```

You can get:

| School | Expr0 | Expr1 | Grouping Id |
| --- | --- | --- | --- |
| Alpha Elementary | Grade 1 | 95 | 0 |
| Alpha Elementary | Grade 2 | 92 | 1 |
| Beta Elementary | Grade 1 | 97 | 0 |
| Beta Elementary | Grade 2 | 93 | 1 |

Although in this case one would normally drop the `Grouping Id` column (but that could be done with an emit).

Reply:
> pivot_longer

WRT to pivot_longer, I think keeping nullification and simply using a subsequent projection with coalesce would be the preferred way to physically implement.

> unreferenced columns/ "like" projection

Hmm. I think of projection as being fundamentally something that doesn't change the number of records so I don't really think of this as the same (fwiw). 

As I think more about it, one of the question I have: can grouping expressions be complex in the systems that use this? If not, this can be come simplified by simply being a int[] "direct projection" column ids + a int[][] of groupings. Then default output order would be direct projection + referenced grouping columns in their original order + group id. If they support arbitrary expressions, we'll need to come up with a more complicated logic around output order (similar to the way it is expressed in aggregate).

Reply:
> WRT to pivot_longer, I think keeping nullification and simply using a subsequent projection with coalesce would be the preferred way to physically implement.

This could work.  I'm against it since I don't think it would be more general, more efficient, or more obvious.  However, I don't want a discussion of pivot_longer to interfere with this PR so we can drop it for later.

> If not, this can be come simplified by simply being a int[] "direct projection" column ids + a int[][] of groupings. Then default output order would be direct projection + referenced grouping columns in their original order + group id

As far as I know, this should work.  I'd support this approach if it works for @chaojun-zhang

Reply:
> I'd like to propose that we use the same pattern as we do in aggregate. The index of the groupings is an additional projection. If this is only duplicating records (and nothing else, that allows us to remove projection_expression and the type.struct output. People can use emit if they want to hide some columns. This is really a physical decomposition of the first half of the logical aggregate with grouping sets so let's be consistent.

@jacques-n  please say [this reply](https://github.com/substrait-io/substrait/pull/368#issuecomment-1469600604).  

This is actually Spark's implementation of the grouping set, spark compiles (optimizes) the grouping set into `Aggregation` on `Expand`.

It is difficult to revert this optimization on the producer side and re-optimizes it on the consumer side.  I did some research for  translating Expand. My research was a bit like disassembling binary codes, which doesn't guarantee that all cases will work. In some cases, because the pattern is fixed, I can work like this. In some cases(i.e. [Spark-9241](https://issues.apache.org/jira/browse/SPARK-9241)), there is no fixed pattern, so I can’t do such a revert.

Reply:
@jacques-n @westonpace 

I am not clear about your discussion. What do we need to do to make progress for his PR now?

Reply:
> I am not clear about your discussion. What do we need to do to make progress for his PR now?

Currently we have:

```
message ExpandRel {
  RelCommon common = 1;
  Rel input = 2;
  Type.Struct output = 3;
  repeated Grouping groupings = 4;
  message Grouping {
    repeated Expression projection_expression = 1;
  }
  substrait.extensions.AdvancedExtension advanced_extension = 10;
}
```

I am proposing (remove `output`):

```
message ExpandRel {
  RelCommon common = 1;
  Rel input = 2;
  repeated Grouping groupings = 4;
  message Grouping {
    repeated Expression projection_expression = 1;
  }
  substrait.extensions.AdvancedExtension advanced_extension = 10;
}
```

I believe that `output` can be calculated from `input` and `groupings` and is redundant.

Jacques is proposing (convert `groupings` from `Expression[][]` to `int[][]`):

```
message ExpandRel {
  RelCommon common = 1;
  Rel input = 2;
  repeated Grouping groupings = 4;
  message Grouping {
    repeated int columns = 1;
  }
  substrait.extensions.AdvancedExtension advanced_extension = 10;
}
```

> As I think more about it, one of the question I have: can grouping expressions be complex in the systems that use this? If not, this can be come simplified by simply being a int[] "direct projection" column ids + a int[][] of groupings. Then default output order would be direct projection + referenced grouping columns in their original order + group id. If they support arbitrary expressions, we'll need to come up with a more complicated logic around output order (similar to the way it is expressed in aggregate).

@jacques-n I have slept on this and I think I disagree.  The way `Expand` is currently implemented in both Acero and Spark is a 2d array of expressions.  Since this is a physical relation I think we should aim to match closely the capability being provided more than the intent.

Also, I said:

> I don't want a discussion of pivot_longer to interfere with this PR so we can drop it for later.

However, I have since discovered that Spark also uses Expand for `pivot_longer` (Spark calls this `UNPIVOT`) so I think it is relevant for both engines.

```
-- !query
SELECT * FROM courseEarnings
UNPIVOT (
  earningsYear FOR year IN (`2012`, `2013`, `2014`)
)
-- !query analysis
Project [course#x, year#x, earningsYear#x]
+- Filter isnotnull(coalesce(earningsYear#x))
   +- Expand [[course#x, 2012, 2012#x], [course#x, 2013, 2013#x], [course#x, 2014, 2014#x]], [course#x, year#x, earningsYear#x]
      +- SubqueryAlias courseearnings
         +- View (`courseEarnings`, [course#x,2012#x,2013#x,2014#x])
            +- Project [cast(course#x as string) AS course#x, cast(2012#x as int) AS 2012#x, cast(2013#x as int) AS 2013#x, cast(2014#x as int) AS 2014#x]
               +- Project [course#x, 2012#x, 2013#x, 2014#x]
                  +- SubqueryAlias courseEarnings
                     +- LocalRelation [course#x, 2012#x, 2013#x, 2014#x]
```

In Acero we require each row (each `Grouping`) to have the same number of expressions.  So the output is fairly straightforward if you think of it like projection (perhaps it would help to think of it as a split into N paths where each path has a project rel and then all the paths are combined in a set union).

I believe Spark also requires each grouping to have the same number of expressions.

Reply:
@westonpace Thank you for your detailed reply！

First,  "unreferenced columns" in your example is actually refered by Spark `Aggregate`'s `AggregateExpression`, so I called it " measure columns". The nullable attribute of the measure column comes from the child nodes of `Expand`, we can **assume** that the nullable attribute of the grouping expression is always true.

```
|unreferenced columns | grouping expressions | grouping id  => your example
|    measure columns  | grouping expressions | grouping id  => 
```

1. With `Type.Struct output`,  we can simply restore the `Expand` operator on the consumer side. 
2. Without `Type.Struct output`,  we need to distinguish between the measure columns and the grouping columns first. For the measure columns, we may need check whether all the projections are the same. For the grouping columns, we can directly set it to null **based on the assumption**.

My question is, adding a field can simplify the work of the consumer side, why do we need to save this field and make the consumer side more complicated? In addition, these assumptions may also be broken in the future.

Reply:
One question: spark generates different `projections `in `Expand `with `grouping sets`, `rollup `and `cube ` expression. And how we handle these different cases if without the output?

Reply:
> My question is, adding a field can simplify the work of the consumer side, why do we need to save this field and make the consumer side more complicated? In addition, these assumptions may also be broken in the future.

I thought this would be very similar to how `ProjectRel` is handled.  For example:

```
ProjectRel {
  expressions = [some_function(selection(3))]
}
```

The output's nullability will be determined by the function `some_function`.  For example, if `some_function` is `abs` (absolute value) then the type and nullability will be the same as column 3.  If `some_function` is `is_null` then the type will be `bool` and the output will be non-null.

In `Expand` I think this is a little harder.  For example, if we have:

```
[
  [field("country"), field("region")]
  [field("country"), literal(null, string())]
  [literal(null, string()), literal(null, string())]
]
```

We have to make sure:

 * The output type must be the same for each position (e.g. `field("country")` and `literal(null, string())` must have the same output type.  `field("region") and `literal(null, string())` must have the same output type).
 * If any output type in a position is nullable then the output type must be nullable.

Putting this together the code might look like:

```
def calculate_output(groupings):
  output_columns=[]
  # Use the first grouping to determine output types
  for expression in groupings[0]:
    output_columns.append(expression.output_type)
  # Make sure the rest of the groupings have the same output types
  # and check to see if they are nullable
  for grouping in groupings[1:]:
    for col_idx in range(len(grouping)):
      # types must be same but nullability can differ
      if output_columns[col_idx].type != grouping[col_idx].output_type:
        raise Exception("All output types for a column must be the same")
      if grouping[col_idx].output_type.nullable:
        output_columns[col_idx].nullable = true
  return output_columns
```

Note: I think the above algorithm works for both "unreferenced columns" and "measure columns".  You don't need to rely on assumption because the the expression `literal(null, string())` will always have a nullable output type so the output for that column will always be nullable.

> why do we need to save this field and make the consumer side more complicated?

The downsides I see are:

 1. It is harder to produce the plan.  The producer now has to calculate the output.
 2. It is possible for a plan to be invalid if output doesn't match the expression (e.g. output is `int` but the expression return type is `bool`).
 3. The plan is larger than it needs to be
 4. This is different than ProjectRel and other relations where we expect the consumer to be able to calculate the output schema.

> One question: spark generates different projections in Expand with grouping sets, rollup and cube expression. And how we handle these different cases if without the output?

I think that is ok.  If you have `ROLLUP` then you have:

```
[
  [field("country"), field("region")],
  [field("country"), literal(null, string())],
  [literal(null, string()), literal(null, string())]
]
```

If you have `CUBE` then you have:

```
[
  [field("country"), field("region")],
  [field("country"), literal(null, string())],
  [literal(null, string()), field("region")],
  [literal(null, string()), literal(null, string())]
]
```

Reply:
@westonpace thanks.
> ```
> def calculate_output(groupings):
> ....
> ```

1. We assume the type should be same at the same posisiton, it would be true now, but I don't think **it would always true**.
2. The cons is all consumers should have the similar codes.

> 4. This is different than ProjectRel and other relations where we expect the consumer to be able to calculate the output schema.

I agree with the 4th. So let's go on with `calculate_output` until we find it can't meet the requirements.

Reply:
@westonpace Thanks for your detailed explanation. I remove the `output `type in `ExpandRel`. Please help to review again.

Reply:
> @westonpace Thanks for your detailed explanation. I remove the `output `type in `ExpandRel`. Please help to review again.

@JkSelf Than you